helix-vcs/src/lib.rs

```diff
@@ -0,0 +1,47 @@
+use std::{collections::HashMap, path::Path};
+
+pub use differ::Differ;
+pub use git::Git;
+
+mod differ;
+mod git;
+mod rope_line_cache;
+
+// TODO: Move to helix_core once we have a generic diff mode
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum LineDiff {
+    Added,
+    Deleted,
+    Modified,
+}
+
+/// Maps line numbers to changes
+pub type LineDiffs = HashMap<usize, LineDiff>;
+
+trait DiffProvider {
+    fn get_file_head(&self, file: &Path) -> Option<Vec<u8>>;
+}
```

Comment:
What is this `file_head`? Why `Vec<u8>` instead of something more specific? Since this is a trait, what is the same thing in other version control?

Reply:
`file_head` is essentially the state of the file that the diff should be computed against. I named it after the `HEAD` in git. Other version control systems call it something different but  I used git terminology because everybody is familiar with that.
However it might indeed be better to choose a different name because I want to allow setting the diff to use a different revision than `HEAD` in the future (or show diffs against a different file). I don't have all the details mapped out of how that is going to work and the `DiffProvider` trait will likely change anyway when I implement that.

Maybe `get_diff_base` might be a better name after all?

The reason this returns a `Vec<u8>` is that the file might not be encoded in UTF-8 (helix supports other encodings after all). The conversion from `Vec<u8>` to something useful (`Rope`) is handeled by helix. I am not 100% sure if this is the right choice (it might be nice to return a `Rope` here if comparing to an open buffer for example). However for the scope of this PR the `Vec<u8>` is the best choice because it avoids duplicating the file decoding code used in helix.

Reply:
Maybe `diff_base`?

Yeah, I am undecided between `Vec<u8>` or `Rope` here so I guess it should be fine.

Reply:
> Maybe `file_base`?

I think `diff_base` fits better because its used in the rest of the implementation in multiple places and conveys more closely what is meant here: a base against which a diff is computed. This will not always be some some base state of the file (and might be a different file altogether in the future). I have updated the name to `get_diff_base` in the lastest commit and also added a bit of documentation. I that's ok for you :)

> Yeah, I am undecided between `Vec<u8>` or Rope here so I guess it should be fine.

I think what will endup workling out better here is best determined in future PRs.
For this PR I will leave in `Vec<u8>` so I don't have to copy a bunch of decoding code that makes the PR harder to review